airbyte-integrations/connectors/source-relational-db/src/test/java/io/airbyte/integrations/source/relationaldb/StateDecoratingIteratorTest.java

```diff
@@ -191,19 +201,27 @@ void testStateEmission1() {
         1);
 
     assertEquals(RECORD_MESSAGE_1, iterator1.next());
-    assertEquals(STATE_MESSAGE_1, iterator1.next());
+    // should emit state 1, but it is unclear whether there will be more
+    // records with the same cursor value, so no state is ready for emission
     assertEquals(RECORD_MESSAGE_2, iterator1.next());
-    assertEquals(STATE_MESSAGE_2, iterator1.next());
+    // emit state 1 because it is the latest state ready for emission
+    assertEquals(STATE_MESSAGE_1, iterator1.next());
     assertEquals(RECORD_MESSAGE_3, iterator1.next());
-    // final state message should only be emitted once
+    assertEquals(STATE_MESSAGE_2, iterator1.next());
+    assertEquals(RECORD_MESSAGE_4, iterator1.next());
     assertEquals(STATE_MESSAGE_3, iterator1.next());
+    assertEquals(RECORD_MESSAGE_5, iterator1.next());
+    // state 3 is not emitted because there is no more record and only
```

Comment:
I'm a bit confused by this comment, based on my understanding every time the iterator executes it _should_ emit a new intermittent `STATE_MESSAGE` that is **not** the initial or the final cursor so in this example I would have expected the `STATE_MESSAGE`s for 2 - 4. How come the `STATE_MESSAGE`s are from 1 - 3 and the code does not emit 4?

Reply:
This stream has no initial cursor (`null`). It is supposed to emit state for every record. Here is what happens.

- Emit record 1.
- It should emit state 1 after record 1. But at this point, it is unclear whether there will be more records with the same cursor. There is no state ready for emission. So no state is emitted.
- Emit record 2. Now it is clear that the cursor has moved to a new value. So state 1 is ready to be emitted.
- Emit state 1.
- Emit record 3. Same logic that state 2 is ready.
- Emit state 2.
- Emit record 4. State 3 is ready.
- Emit state 3.
- Emit record 5. State 4 is ready.
- There is no more record. So only the final state, state 5, will be emitted. This is why state 4 is NOT emitted.